### PR TITLE
[test] fix for hasPrefix and hasSuffix test

### PR DESCRIPTION
**Pull request by glessard**
The previous version wrongly cached a result, which in turn masked a bug in the bridged `hasSuffix()`.
(reported as https://bugs.swift.org/browse/SR-243)

This version also trips `hasPrefix`, but that involves `\u{0}`; the correct behaviour there is less obvious.

### CODE REVIEW

**Comment by glessard** [comment]
This bug fix surfaces two cases that always failed, but were masked.
They are now both annotated with xfail and links to bug reports.

**Comment by glessard** [comment]
Rebased after the changes in https://github.com/apple/swift/pull/1131

**Comment by tkremenek** [comment]
@swift-ci test

**Comment by tkremenek** [comment]
@gribozavr please review

**Comment by glessard** [comment]
@tkremenek The swift-ci failure was unrelated to these changes; test again?

**Comment by jrose-apple** [comment]
Drive-by test restarting:

@swift-ci Please test OS X platform

**Comment by jckarter** [comment]
@gribozavr This ok to merge?

**Comment by gribozavr** [comment]
@glessard Sorry, I don't understand what this patch changes -- it looks like a pure refactoring to me.  I see that you are removing the xfail predicates from the test list, but you're adding them back with the map() function.  Why is this making a difference?  I'm probably missing something, I would be glad if you could explain to me.

**Comment by glessard** [comment]
The bug fix is https://github.com/apple/swift/pull/629/commits/21d5da7a768c1a36089525de2f9b4b4cbff86215, the rest is dealing with the consequences of it (and of https://github.com/apple/swift/pull/1131, since they have distinct sets of xfails).

In the original version of 'checkHasPrefixHasSuffix', a pair of expected results is calculated, and then are used twice each. This wasn't correct. In particular, if the first character of `lhs` is combining, `("abc"+lhs).hasSuffix(rhs)` should not be equal to `lhs.hasSuffix(rhs)`, yet the test expected that to be true (see https://github.com/apple/swift/pull/629/commits/21d5da7a768c1a36089525de2f9b4b4cbff86215#diff-f7824440e4e317d38fc0554573876bccL261). The original test passed the `("\u{0301}", "\u{0341}")` case by matching a bug of NSString/CFString (https://bugs.swift.org/browse/SR-243). Without that NSString bug, the original formulation of the test wouldn't have passed.

The fix simply ensures the expected result is recalculated for every single use, by moving the extra suffix and prefix tests to a separate call of `checkHasPrefixHasSuffix`. (I stumbled upon this while porting the tests for https://github.com/apple/swift/pull/440)

**Comment by glessard** [comment]
@gribozavr Did the above response clarify the patch?

File: test/1_stdlib/StringAPI.swift [hunk 1]
```diff
@@ -24,34 +24,44 @@ struct ComparisonTest {
...
   }
+
+  func replacingPredicate(_ xfail: TestRunPredicate) -> ComparisonTest {
+    return ComparisonTest(expectedUnicodeCollation, lhs: lhs, rhs: rhs,
+      xfail: xfail, loc: loc)
+  }
 }
 
-let comparisonTests = [
+let tests = [
```
Comment (CodaFi) [inline test/1_stdlib/StringAPI.swift]: It doesn't seem like anything else in the file is using `tests`.  Why not just fold the results of the map into the test itself?

**Comment by CodaFi** [comment]
Let's try to revive this:

@swift-ci please smoke test macOS platform.

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

File: test/1_stdlib/StringAPI.swift [hunk 2]
```diff
@@ -24,34 +24,44 @@ struct ComparisonTest {
...
   ) {
+    self = ComparisonTest(expectedUnicodeCollation, lhs: lhs, rhs: rhs,
+      xfail: xfail, loc: SourceLoc(file, line, comment: "test data"))
+  }
+
+  init(
+    _ expectedUnicodeCollation: ExpectedComparisonResult,
+    lhs: String, rhs: String,
+    xfail: TestRunPredicate, loc: SourceLoc
+  ) {
```
Comment (CodaFi) [inline test/1_stdlib/StringAPI.swift]: Now that this is just a change to a default argument, this initializer isn't needed.

File: test/1_stdlib/StringAPI.swift (hunk 1 above)
Comment (glessard) [inline test/1_stdlib/StringAPI.swift]: The tests for comparisons and substrings (prefix and suffix) don't overlap cleanly; this is why I separated them (map at line 147, map at line 290).

File: test/1_stdlib/StringAPI.swift (hunk 2 above)
Comment (glessard) [inline test/1_stdlib/StringAPI.swift]: Agreed.

File: test/1_stdlib/StringAPI.swift (hunk 1 above)
Comment (CodaFi) [inline test/1_stdlib/StringAPI.swift]: There definitely needs to be a little bit of documentation on both of the maps and on the special cases of the literal here explaining just that.  It's hard to see where and why things XFail without it.

**Comment by CodaFi** [comment]
I think I'm out of nitpicks if this was pure refactoring.  Thank you @glessard.

**Comment by CodaFi** [comment]
Once you've added a bit of docs, please squash this down to a patch or two.

**Comment by glessard** [comment]
@CodaFi I think this should do it; thanks for the feedback

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
[macOS](https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx-smoke-test/1348/)
[Linux](https://ci.swift.org/view/Pull%20Request/job/swift-PR-Linux-smoke-test/837/)

// @shahmishal

**Comment by shahmishal** [comment]
This issue looks like: 
https://github.com/janinko/ghprb/issues/150
https://github.com/janinko/ghprb/issues/284

going to retrigger the build to see if it picks up the new sha. 

@swift-ci Please smoke test

**Comment by CodaFi** [comment]
I think we're good to go here.  Merging.

**Comment by jrose-apple** [comment]
Ah, @gribozavr, did you have any remaining comments?

**Comment by gribozavr** [comment]
LGTM, thanks!

**Pull request merged**